From ceca12907741ddbb6ca4e845ee914f9cfb165152 Mon Sep 17 00:00:00 2001
From: Timple Raj M <quic_timple@quicinc.com>
Date: Mon, 17 Jul 2023 16:17:04 +0530
Subject: [PATCH] samples: qmi: Update the qmi sample client driver

This patch moves "qmi_sample_client driver" specific data
structures from header file include/linux/soc/qcom/qmi.h,
since it is required for internal validation only.

For next kernel upgrade, just copy qmi sample client driver alone
i.e no need to touch any upstream file.

Change-Id: I67f944531bdd417020ad90a40a050c398dabd9d2
Signed-off-by: Timple Raj M <quic_timple@quicinc.com>
---
 include/linux/soc/qcom/qmi.h    |  3 --
 samples/qmi/qmi_sample_client.c | 73 ++++++++++++++++++---------------
 2 files changed, 40 insertions(+), 36 deletions(-)

diff --git a/include/linux/soc/qcom/qmi.h b/include/linux/soc/qcom/qmi.h
index e15001efaff9..469e02d2aa0d 100644
--- a/include/linux/soc/qcom/qmi.h
+++ b/include/linux/soc/qcom/qmi.h
@@ -235,9 +235,6 @@ struct qmi_handle {
 	struct mutex txn_lock;
 
 	const struct qmi_msg_handler *handlers;
-	struct list_head data_list;
-	struct completion complete;
-	atomic_t cnt, async_cnt, async_rsp, async_req, pass, fail;
 };
 
 int qmi_add_lookup(struct qmi_handle *qmi, unsigned int service,
diff --git a/samples/qmi/qmi_sample_client.c b/samples/qmi/qmi_sample_client.c
index 39c591db1bd3..dd7e3ab427d9 100644
--- a/samples/qmi/qmi_sample_client.c
+++ b/samples/qmi/qmi_sample_client.c
@@ -511,6 +511,14 @@ static const struct qmi_elem_info test_data_resp_msg_v01_ei[] = {
 	{}
 };
 
+struct qmi_sample_pvt_data {
+	struct list_head data_list;
+	struct completion complete;
+	atomic_t cnt, pass, fail;
+};
+
+static struct qmi_sample_pvt_data *pvt_data;
+
 /*
  * ping_write() - ping_pong debugfs file write handler
  * @file:	debugfs file context
@@ -1109,7 +1117,7 @@ static struct dentry *qmi_debug_dir;
 static void update_status(struct qmi_handle *qmi)
 {
 	unsigned int max = nthreads * niterations;
-	unsigned int count = atomic_read(&qmi->cnt);
+	unsigned int count = atomic_read(&pvt_data->cnt);
 	unsigned int percent;
 	static unsigned int pre_percent;
 
@@ -1128,7 +1136,7 @@ static int test_qmi_ping_pong_send_msg(struct qmi_handle *qmi)
 	struct qmi_txn txn;
 	int ret;
 
-	atomic_inc(&qmi->cnt);
+	atomic_inc(&pvt_data->cnt);
 
 	memcpy(req.ping, "ping", sizeof(req.ping));
 
@@ -1141,7 +1149,7 @@ static int test_qmi_ping_pong_send_msg(struct qmi_handle *qmi)
 			TEST_PING_REQ_MAX_MSG_LEN_V01,
 			test_ping_req_msg_v01_ei, &req);
 	if (ret < 0) {
-		atomic_inc(&qmi->fail);
+		atomic_inc(&pvt_data->fail);
 		qmi_txn_cancel(&txn);
 		return ret;
 	}
@@ -1149,11 +1157,11 @@ static int test_qmi_ping_pong_send_msg(struct qmi_handle *qmi)
 	ret = qmi_txn_wait(&txn, 5 * HZ);
 	if (ret < 0) {
 		pr_err("Failed to get response on the txn\n");
-		atomic_inc(&qmi->fail);
+		atomic_inc(&pvt_data->fail);
 		return ret;
 	}
 
-	atomic_inc(&qmi->pass);
+	atomic_inc(&pvt_data->pass);
 	mutex_lock(&status_print_lock);
 	update_status(qmi);
 	mutex_unlock(&status_print_lock);
@@ -1168,18 +1176,18 @@ static int test_qmi_data_send_msg(struct qmi_handle *qmi, unsigned int data_len)
 	struct qmi_txn txn;
 	int ret, i;
 
-	atomic_inc(&qmi->cnt);
+	atomic_inc(&pvt_data->cnt);
 
 	req = kzalloc(sizeof(*req), GFP_KERNEL);
 	if (!req) {
-		atomic_inc(&qmi->fail);
+		atomic_inc(&pvt_data->fail);
 		return -ENOMEM;
 	}
 
 	resp = kzalloc(sizeof(*resp), GFP_KERNEL);
 	if (!resp) {
 		kfree(req);
-		atomic_inc(&qmi->fail);
+		atomic_inc(&pvt_data->fail);
 		return -ENOMEM;
 	}
 
@@ -1190,7 +1198,7 @@ static int test_qmi_data_send_msg(struct qmi_handle *qmi, unsigned int data_len)
 
 	ret = qmi_txn_init(qmi, &txn, test_data_resp_msg_v01_ei, resp);
 	if (ret < 0) {
-		atomic_inc(&qmi->fail);
+		atomic_inc(&pvt_data->fail);
 		goto out;
 	}
 
@@ -1200,14 +1208,14 @@ static int test_qmi_data_send_msg(struct qmi_handle *qmi, unsigned int data_len)
 			test_data_req_msg_v01_ei, req);
 	if (ret < 0) {
 		qmi_txn_cancel(&txn);
-		atomic_inc(&qmi->fail);
+		atomic_inc(&pvt_data->fail);
 		goto out;
 	}
 
 	ret = qmi_txn_wait(&txn, 5 * HZ);
 
 	if (ret < 0) {
-		atomic_inc(&qmi->fail);
+		atomic_inc(&pvt_data->fail);
 		goto out;
 	}
 
@@ -1215,12 +1223,12 @@ static int test_qmi_data_send_msg(struct qmi_handle *qmi, unsigned int data_len)
 			resp->data_len != req->data_len ||
 			memcmp(resp->data, req->data, req->data_len)) {
 		pr_err("response data doesn't match expectation\n");
-		atomic_inc(&qmi->fail);
+		atomic_inc(&pvt_data->fail);
 		ret = -EINVAL;
 		goto out;
 	} else {
 		pr_debug("Data valid\n");
-		atomic_inc(&qmi->pass);
+		atomic_inc(&pvt_data->pass);
 	}
 
 	mutex_lock(&status_print_lock);
@@ -1238,12 +1246,10 @@ int qmi_process_user_input(void *data)
 {
 	struct test_qmi_data *qmi_data, *temp_qmi_data;
 	unsigned short index = 0;
-	struct qmi_handle *qmi;
-	struct list_head *data_list = data;
+	struct qmi_handle *qmi = data;
+	struct list_head *data_list = &pvt_data->data_list;
 
-	qmi = container_of(data, struct qmi_handle, data_list);
-
-	wait_for_completion_timeout(&qmi->complete, msecs_to_jiffies(1000));
+	wait_for_completion_timeout(&pvt_data->complete, msecs_to_jiffies(1000));
 
 	list_for_each_entry_safe(qmi_data, temp_qmi_data, data_list, list) {
 		atomic_inc(&qmi_data->refs_count);
@@ -1265,7 +1271,7 @@ int qmi_process_user_input(void *data)
 
 		if (atomic_dec_and_test(&qmi_data->refs_count)) {
 			pr_info("Test Completed. Pass: %d Fail: %d\n",
-					atomic_read(&qmi->pass), atomic_read(&qmi->fail));
+					atomic_read(&pvt_data->pass), atomic_read(&pvt_data->fail));
 			list_del(&qmi_data->list);
 			kfree(qmi_data);
 			qmi_data = NULL;
@@ -1296,19 +1302,16 @@ static int test_qmi_open(struct inode *ip, struct file *fp)
 	pr_info("Total commands: %lu (Threads: %lu Iteration: %lu)\n",
 			nthreads * niterations, nthreads, niterations);
 
-	atomic_set(&qmi->cnt, 0);
-	atomic_set(&qmi->pass, 0);
-	atomic_set(&qmi->fail, 0);
-	atomic_set(&qmi->async_cnt, 0);
-	atomic_set(&qmi->async_req, 0);
-	atomic_set(&qmi->async_rsp, 0);
+	atomic_set(&pvt_data->cnt, 0);
+	atomic_set(&pvt_data->pass, 0);
+	atomic_set(&pvt_data->fail, 0);
 
-	init_completion(&qmi->complete);
+	init_completion(&pvt_data->complete);
 	for (index = 0; index < nthreads; index++) {
 		snprintf(thread_name, sizeof(thread_name), "qmi_""%s_%d",
 				fp->f_path.dentry->d_parent->d_iname, index);
 		qmi_task = kthread_run(qmi_process_user_input,
-					&qmi->data_list, thread_name);
+					qmi, thread_name);
 	}
 	return 0;
 
@@ -1377,8 +1380,8 @@ static ssize_t test_qmi_write(struct file *fp, const char __user *buf,
 	atomic_set(&qmi_data->refs_count, 0);
 
 	qmi = fp->private_data;
-	list_add_tail(&qmi_data->list, &qmi->data_list);
-	complete_all(&qmi->complete);
+	list_add_tail(&qmi_data->list, &pvt_data->data_list);
+	complete_all(&pvt_data->complete);
 
 	return count;
 }
@@ -1397,6 +1400,10 @@ static int qmi_sample_probe(struct platform_device *pdev)
 	char path[20];
 	int ret;
 
+	pvt_data = devm_kzalloc(&pdev->dev, sizeof(*pvt_data), GFP_KERNEL);
+	if (!pvt_data)
+		return -ENOMEM;
+
 	sample = devm_kzalloc(&pdev->dev, sizeof(*sample), GFP_KERNEL);
 	if (!sample)
 		return -ENOMEM;
@@ -1417,7 +1424,7 @@ static int qmi_sample_probe(struct platform_device *pdev)
 
 	snprintf(path, sizeof(path), "%d:%d", sq->sq_node, sq->sq_port);
 
-	INIT_LIST_HEAD(&sample->qmi.data_list);
+	INIT_LIST_HEAD(&pvt_data->data_list);
 	sample->de_dir = debugfs_create_dir(path, qmi_debug_dir);
 	if (IS_ERR(sample->de_dir)) {
 		ret = PTR_ERR(sample->de_dir);
@@ -1563,7 +1570,6 @@ static int qmi_sample_probe(struct platform_device *pdev)
 static int qmi_sample_remove(struct platform_device *pdev)
 {
 	struct qmi_sample *sample = platform_get_drvdata(pdev);
-	struct qmi_handle *qmi = &sample->qmi;
 	struct test_qmi_data *qmi_data, *temp_qmi_data;
 
 	debugfs_remove(sample->de_rail_voltage_uv);
@@ -1583,11 +1589,12 @@ static int qmi_sample_remove(struct platform_device *pdev)
 
 	qmi_handle_release(&sample->qmi);
 
-	list_for_each_entry_safe(qmi_data, temp_qmi_data, &qmi->data_list, list) {
+	list_for_each_entry_safe(qmi_data, temp_qmi_data, &pvt_data->data_list, list) {
 		list_del(&qmi_data->list);
 		kfree(qmi_data);
+		qmi_data = NULL;
 	}
-	list_del(&qmi->data_list);
+	list_del(&pvt_data->data_list);
 
 	return 0;
 }
-- 
2.34.1

